sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,318 @@
+
+# Example: Simple Naming System
+
+To demonstrate the expressiveness of this smart contracting language,
+let's look at an example smart contract which implements a simple
+naming system with just two kinds of transactions: _preorder_
+and _register_. The requirements of the system are as follows:
+
+1. Names may only be owned by one principal
+2. A register is only allowed if there is a corresponding preorder
+   with a matching hash
+3. A register transaction must be signed by the same principal who
+   paid for the preorder
+4. A preorder must have paid at least the price of the name. Names
+   are represented as integers, and any name less than 100000 costs
+   1000 microstacks, while all other names cost 100 microstacks.
+5. Preorder hashs are _globally_ unique.
+
+In this simple scheme, names are represented by integers, but in
+practice, a buffer would probably be used.
+
+```scheme
+(define burn-address '1111111111111111111114oLvT2)
+(define (price-function name)
+  (if (< name 1e5) 1000 100))
+
+(defmap name-map 
+  ((name integer)) ((buyer principal)))
+(defmap preorder-map
+  ((name-hash (buffer 160)))
+  ((buyer principal) (paid integer)))
+
+(define (tx-preorder 
+           (buyer-principal principal)
+           (name-hash (buffer 20))
+           (name-price integer))
+  (if (stacks-transfer!
+        buyer-principal name-price burn-address name-price)
+      (insert-entry! preorder-map
+        (tuple #name-hash name-hash)
+        (tuple #paid name-price
+               #buyer buyer-principal))
+      false))
+
+(define (tx-register 
+           (buyer-principal principal)
+           (recipient-principal principal)
+           (name integer)
+           (salt integer)
+  (if (signed-by? buyer-principal)
+    (let ((preorder-entry
+            (fetch-entry preorder-map
+                         (tuple #name-hash (hash160 name salt))))
+          (name-entry 
+            (fetch-entry name-map (tuple #name name))))
+      (if (and
+           ;; must be preordered
+           (not (eq? preorder-entry) 'null)
+           ;; name shouldn't *already* exist
+           (eq? name-entry 'null)
+           ;; preorder must have paid enough
+           (<= (price-funcion name) 
+               (get #paid preorder-entry))
+           ;; preorder must have been buyer-principle
+           (eq? buyer-principle
+                (get #buyer preorder-entry)))
+          (begin
+            (insert-entry! name-table
+              (tuple #name name)
+              (tuple #owner recipient)))
+          false))
+    false)))
+```
+
+
+Note that Blockstack PBC intends to supply a full BNS (Blockstack
+Naming System) smart contract, as well as formal proofs that certain
+desirable properties hold (e.g. "names are globally unique", "a
+revoked name cannot be updated or transferred", "names cost stacks
+based on their namespace price function", "only the principal can
+reveal a name on registration", etc.).
```

Comment:
Things that still need to be added:

* how do we invoke methods?
    * how do we do it from Blockstack Core?
    * how do we call methods from transactions?
* no dynamic dispatch *yet*
* static analysis is done eagerly on its methods when it's uploaded
* no upgrade path for smart contract -- turing-incompleteness helps remove the need for upgrades
* no call cycles -- a smart contract can't re-enter itself
* smart contract inputs?
    * principal (this is a primitive type)
    * method name
    * method arguments
    * this contract (txid)
* smart contract outputs?
    * log of writes with a materialized view of the state of the maps
    * log of affected accounts (for token transfers)
* how do I do accounting?
* addressing scheme is by tx hash
    * there is a "store contract" transaction (e.g. SCxxxxx); the txid is used to address its methods
* smart contracts own tokens
    * there needs to be primitives for sending tokens out of the smart contract
        * "getBalance(tokenType)" and "sendTokens(tokenType, units)"
    * smart contract token address is a differently-versioned hash160 of its txid
        * this lets us send tokens to a smart contract
    * if the smart contract spends tokens, it appends to the account balances table with a back-pointer to the tx that caused it to spend
* every smart contract method *must* return "true" or "false"
* smart contracts, when invoked, return "true" or "false"
    * if "true", all changes are made persistent
    * if "false", no changes are made persistent
* limits on writes to tables?
    * limit number of writes (e.g. 1024)
* tx fee is known beforehand from static analysis
* develop cost metric and fuzz the shit out of it (this is a different SIP)

Reply:
> smart contracts own tokens

Yes, contracts should have a balance, but wouldn't this always be in Stacks? Then, why the `tokenType` argument?

Reply:
Yeah, let's drop this for now.  I was getting ahead of myself.

Reply:
the latest version addressed those bullet points.

notable differences from the bullet points above:

1. contracts owning tokens -> this is realized by contracts having a 'contract identifier' which is an object of the principal type. Contracts have a function `(contract-send-tx (..))` which executes its arguments with the `tx-sender` variable set to the _contract identifier_ rather than the original _transaction sender_. This allows not only contracts to own tokens, but anything else that uses principals to represent owners.

2. smart contract outputs -> rather than output a log of operations, the smart contract VM should have a connection to a database. The database itself would track the log of operations.

Reply:
Ping @jcnelson !

Reply:
Yup, we're good here.  I've added other comments elsewhere.